lambda/translcomprehension.ml

```diff
@@ -2,304 +2,406 @@ open Lambda
+  Llet(let_kind, value_kind, var, init, body)
 
-let make_array size init ~loc=
-  let prim_make_arr =
+let gen_bindings bindings body =
+  List.fold_right gen_binding bindings body
+
+(* Translate a clause into some initialising bindings, a variable
+   that will be bound to the number of iterations in the clause by
+   those bindings, and lambda code that performs the iterations. *)
+let transl_arr_clause ~transl_exp ~scopes ~loc clause body =
+  let len_var = Ident.create_local "len_var" in
+  let bindings, for_ =
+    match clause with
+    | In (pat , e2) ->
+        let in_var = Ident.create_local "in_var" in
+        let in_kind = Typeopt.array_kind e2 in
+        let in_binding =
+          binding Strict Pgenval in_var (transl_exp ~scopes e2)
+        in
+        let len_binding =
+          let init = Lprim( (Parraylength(in_kind)), [Lvar(in_var)], loc) in
+          binding Alias Pintval len_var init
+        in
+        let index = Ident.create_local "index" in
+        let for_ =
+          Lfor(index, (int 0), Lprim(Psubint, [Lvar(len_var); int 1], loc) , Upto,
+            Matching.for_let ~scopes pat.pat_loc
+              (Lprim(Parrayrefu(in_kind),
+                     [Lvar(in_var); Lvar(index)], loc)) pat body)
+        in
+        [in_binding; len_binding], for_
+    | From_to(id, _, e2, e3, dir) ->
+        let from_var = Ident.create_local "from" in
+        let from_binding =
+          binding Strict Pintval from_var (transl_exp ~scopes e2)
+        in
+        let to_var = Ident.create_local "to" in
+        let to_binding =
+          binding Strict Pintval to_var (transl_exp ~scopes e3)
+        in
+        let low, high =
+          match dir with
+          | Upto -> Lvar from_var, Lvar to_var
+          | Downto -> Lvar to_var, Lvar from_var
+        in
+        let len_binding =
+          let init =
+            Lprim(Psubint, [Lprim(Paddint, [high; int 1], loc); low], loc)
+          in
+          binding Alias Pintval len_var init
+        in
+        let for_ = Lfor(id, Lvar from_var, Lvar to_var, dir, body) in
+        [from_binding; to_binding; len_binding], for_
+  in
+  bindings, len_var, for_
+
+(* Generate code to iterate over a comprehension block, along with some
+   initialising bindings.  The bindings will also bind the given
+   [length_var] ident to the total number2 of iterations in the
+   block. *)
+let iterate_arr_block ~transl_exp ~loc ~scopes
+      {clauses; guard} length_var body =
+  let body =
+    match guard with
+    | None -> body
+    | Some guard ->
+      Lifthenelse(transl_exp ~scopes guard, body, lambda_unit)
+  in
+  let body, lengths, rev_bindings =
+    List.fold_left
+      (fun (body, lengths, rev_bindings) clause ->
+         let new_bindings, new_length_var, body =
+           transl_arr_clause ~transl_exp ~scopes ~loc clause body
+         in
+         let rev_bindings = List.rev_append new_bindings rev_bindings in
+         let lengths = new_length_var :: lengths in
```

Comment:
We could directly do the "Some/ None" magic here to multiply the the lengths. So we do not have to fold left twice.

Reply:
Good idea. Done.